File: backend/app/workers/media_kind.py
"""Media category inference utility.

Maps file extensions to the four UX categories used by Results drawer
and cross-tool routing. Keep this aligned with
`core/frontend/src/utils/mediaType.ts` (frontend mirror).
"""
from __future__ import annotations

from enum import Enum
from pathlib import Path
from typing import Optional


class MediaKind(str, Enum):
    IMAGE = "image"
    AUDIO = "audio"
    VIDEO = "video"
    DOCUMENT = "document"


_IMAGE_EXTS = {".jpg", ".jpeg", ".png", ".gif", ".webp", ".bmp", ".tiff", ".tif", ".svg", ".ico", ".avif", ".heic", ".heif"}
_AUDIO_EXTS = {".mp3", ".wav", ".flac", ".ogg", ".aac", ".m4a", ".wma", ".opus", ".mid", ".midi"}
_VIDEO_EXTS = {".mp4", ".mov", ".webm", ".avi", ".mkv", ".flv", ".wmv", ".m4v"}
_DOCUMENT_EXTS = {".pdf", ".doc", ".docx", ".txt", ".srt", ".vtt", ".md", ".csv", ".json", ".html", ".odt"}


def infer_kind(filename: str) -> Optional[MediaKind]:
    """Return the MediaKind for a filename's extension, or None if unknown."""
    ext = Path(filename).suffix.lower()
    if ext in _IMAGE_EXTS:
        return MediaKind.IMAGE
    if ext in _AUDIO_EXTS:
        return MediaKind.AUDIO
    if ext in _VIDEO_EXTS:
        return MediaKind.VIDEO
    if ext in _DOCUMENT_EXTS:
        return MediaKind.DOCUMENT
    return None

File: frontend/src/utils/mediaKind.ts
/**
 * 五官媒體歸類(pipeline/results 用)——與後端 backend/app/workers/media_kind.py
 * 完全鏡像,由 __tests__/mediaKind.test.ts 的後端表快照釘住。
 * 注意:這與 mediaType.ts 的 ToolType(4 視圖路由)是兩套分類學,不要混用。
 */
export type MediaKindT = 'video' | 'audio' | 'image' | 'document'

export const MEDIA_KIND_EXTS: Record<MediaKindT, Set<string>> = {
  image: new Set(['jpg', 'jpeg', 'png', 'apng', 'gif', 'webp', 'bmp', 'tiff', 'tif', 'svg', 'ico', 'avif', 'heic', 'heif']),
  audio: new Set(['mp3', 'wav', 'flac', 'ogg', 'aac', 'm4a', 'wma', 'opus', 'mid', 'midi']),
  video: new Set(['mp4', 'mov', 'webm', 'avi', 'mkv', 'flv', 'wmv', 'm4v']),
  document: new Set(['pdf', 'doc', 'docx', 'txt', 'srt', 'vtt', 'md', 'csv', 'json', 'html', 'odt', 'lrc']),
}

export function detectMediaKind(filename: string): MediaKindT | null {
  const dot = filename.lastIndexOf('.')
  if (dot < 0 || dot === filename.length - 1) return null
  const ext = filename.slice(dot + 1).toLowerCase()
  for (const kind of Object.keys(MEDIA_KIND_EXTS) as MediaKindT[]) {
    if (MEDIA_KIND_EXTS[kind].has(ext)) return kind
  }
  return null
}
